File: 35. Search Insert Position/Searchinsert.c
int binarySearch(int* nums, int start, int end, int target) {
    // Calculate the middle index
    int mid = start + (end - start) / 2;
    
    // Base case: If the search range is invalid, return index(which is the position of the target if it was to be found)
    if (start > end)
        return start;


    // If the middle element is the target, return its index
    if (nums[mid] == target)
        return mid;

    // If the middle element is less than the target, search the right half
    if (nums[mid] < target)
        return binarySearch(nums, mid + 1, end, target);

    // If the middle element is greater than the target, search the left half
    if (nums[mid] > target)
        return binarySearch(nums, start, mid - 1, target);
    return start;
}

int searchInsert(int* nums, int numsSize, int target) {
    // Call the binarySearch helper function with initial parameters
    return binarySearch(nums, 0, numsSize - 1, target);
}

File: 35. Search Insert Position/SearchInsert.py
def searchInsert(nums: list, target: int) -> int:
    # Initialize the start and end indices of the search range
    start, end = 0, len(nums) - 1
    
    # Calculate the initial midpoint
    mid = (end) // 2

    # Perform binary search while the search range is valid
    while start <= end:
        # If the middle element is less than the target, adjust the start index and recalculate the midpoint
        if nums[mid] < target:
            start = mid + 1
            length = end - start
            mid = start + (length // 2)
        
        # If the middle element is greater than the target, adjust the end index and recalculate the midpoint
        elif nums[mid] > target:
            end = mid - 1
            length = end - start
            mid = start + (length // 2)
        
        # If the middle element is equal to the target, return the index
        else:
            return mid

    # If the target element is not found, return index where it should be
    return start
